rpcs3/rpcs3qt/debugger_list.cpp

```diff
@@ -34,25 +34,77 @@ debugger_list::debugger_list(QWidget* parent, std::shared_ptr<gui_settings> gui_
 
 	setSizeAdjustPolicy(QListWidget::AdjustToContents);
 	setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+
+	connect(this, &QListWidget::currentRowChanged, this, [this](int row)
+	{
+		if (row < 0)
+		{
+			selected_instruction = -1;
+			return;
+		}
+
+		u32 pc = m_start_addr;
+
+		while (m_cpu && m_cpu->id_type() == 0x55 && row)
+		{
+			// If scrolling forwards (downwards), we can skip entire commands
+			// Backwards is impossible though
+			pc += std::max<u32>(m_disasm->disasm(pc), 4);
+			row--;
+		}
+
+		selected_instruction = pc + row * 4;
+	});
 }
 
 void debugger_list::UpdateCPUData(cpu_thread* cpu, CPUDisAsm* disasm)
 {
 	if (m_cpu != cpu)
 	{
 		m_cpu = cpu;
-		m_selected_instruction = -1;
+		selected_instruction = -1;
+		showing_selected_instruction = false;
 	}
 
 	m_disasm = disasm;
 }
 
-u32 debugger_list::GetCenteredAddress(u32 address) const
+u32 debugger_list::GetStartAddress(u32 address)
 {
-	return address - ((m_item_count / 2) * 4);
+	const u32 steps = m_item_count / 3;
+
+	u32 result = address;
+
+	if (m_cpu && m_cpu->id_type() == 0x55)
+	{
+		if (auto [count, res] = static_cast<rsx::thread*>(m_cpu)->try_get_pc_of_x_cmds_backwards(steps, address); count == steps)
+		{
+			result = res;
+		}
+	}
+	else
+	{
+		result = address - (steps * 4);
+	}
+
+	if (m_start_addr > address)
+	{
+		m_pc = address;
+		m_start_addr = result;
+		return m_start_addr;
+	}
+
+	if (address > m_pc)
+	{
+		m_pc = address;
+		m_start_addr = result;
+		return m_start_addr;
+	}
```

Comment:
almost identical code can be merged.
return statements are redundant

Reply:
I removed identical code, the return value can remain for convenience.